### PR TITLE
[Python] Re-work Python bindings using upstream improvements.

### PR DESCRIPTION
**Pull request by mikeurbach**
This is mostly mechanical for us. Some notable changes:

* Use a unified build in Python docs and CI
* New CMake defaults for some variables in CMakeLists.txt
* Use the new CMake functions for Python bindings and PyCDE
* Update imports for generated Python and Python extension libraries
* Update PYTHONPATH to reflect the new location under unified builds

### CODE REVIEW

File: docs/PythonBindings.md [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
 ```
-# in the LLVM/MLIR build directory
-cmake [...] -DMLIR_ENABLE_BINDINGS_PYTHON=ON
-
-# in the CIRCT build directory
-cmake [...] -DCIRCT_BINDINGS_PYTHON_ENABLED=ON
+$ cd circt
+$ mkdir llvm/build
+$ cd llvm/build
+$ cmake -G Ninja ../llvm \
```
Comment (teqdruid) [inline docs/PythonBindings.md:9]: I think this is missing:

```
            -DLLVM_EXTERNAL_PROJECTS=circt \
            -DLLVM_EXTERNAL_CIRCT_SOURCE_DIR=.. \
```

Comment (mikeurbach) [reply docs/PythonBindings.md:9]: yep thanks

**Comment by teqdruid** [comment]
What are you observing in terms of compile times? I've been avoiding unified builds (both in the integration tests and locally) since I assumed they would be slower. Presumably, I'll need to switch over locally as well.

**Comment by mikeurbach** [comment]
@teqdruid I'm going to answer this in the forum, since it seems more relevant to the broader discussion about unified builds, and I don't want it to get buried in some PR.

**Comment by teqdruid** [comment]
Just saw that forum post.

**Comment by stellaraccident** [comment]
FYI - here is a CI config I'm testing now for npcomp (cribbed somewhat from mlir-hs and the pre-existing LLVM install-cache approach).

```
name: Build and Test

#on: [push, pull_request]
on: [push]

jobs:
  build:
    name: Build and Test (Release Asserts)
    runs-on: ubuntu-20.04
    steps:
    - name: Get npcomp
      uses: actions/checkout@v2
      with:
        submodules: 'true'
    - name: Install python depends
      run: |
        python3 -m pip install -r $GITHUB_WORKSPACE/external/llvm-project/mlir/python/requirements.txt
    - name: Install pytorch_nightly depends
      run: |
        python3 -m pip install --pre 'torch==1.10.0.dev20210630+cpu' 'torchvision==0.11.0.dev20210630+cpu' -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
    - name: Install Ninja
      uses: llvm/actions/install-ninja@55d844821959226fab4911f96f37071c1d4c3268
    - name: Get Submodule Hash
      id: get-submodule-hash
      run: echo "::set-output name=hash::$(md5sum $(git submodule status))"
      shell: bash
    - name: Ccache for C++ compilation
      uses: hendrikmuhs/ccache-action@4687d037e4d7cf725512d9b819137a3af34d39b3
      with:
        key: ${{ runner.os }}-${{ steps.get-submodule-hash.outputs.hash }}
    - name: Build and Test npcomp (Assert)
      run: |
        cd $GITHUB_WORKSPACE
        mkdir build
        cd build
        cmake .. -GNinja \
          -DCMAKE_BUILD_TYPE=Release \
          -DCMAKE_LINKER=lld \
          -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
          -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ \
          -DPython3_EXECUTABLE=/usr/bin/python3 \
          -DLLVM_ENABLE_ASSERTIONS=ON \
          -DLLVM_TARGETS_TO_BUILD=host \
          -DNPCOMP_ENABLE_PYTORCH=ON
        ninja
        ninja check-npcomp check-frontends-pytorch
```

**Review by stellaraccident — APPROVED** [review]

**Comment by mikeurbach** [comment]
Thanks @stellaraccident, I'm working on the same idea over here. Since we're using a Docker container, I'm going to need to install `ccache` manually, but otherwise something like that should work well for us. Is it important to have the submodule commit in the `ccache` key? As I've been using it locally, I've just trusted it to know when LLVM has changed and it has seemed to just work so far. Am I missing something? I was planning to include the various pieces of our build matrix in the key (release/debug, etc.).

**Comment by stellaraccident** [comment]
> Thanks @stellaraccident, I'm working on the same idea over here. Since we're using a Docker container, I'm going to need to install `ccache` manually, but otherwise something like that should work well for us. Is it important to have the submodule commit in the `ccache` key? As I've been using it locally, I've just trusted it to know when LLVM has changed and it has seemed to just work so far. Am I missing something? I was planning to include the various pieces of our build matrix in the key (release/debug, etc.).

I dunno. My thinking was that it is common for there to be PRs in flight for multiple versions of LLVM and it would be good if they didn't thrash each other.

**Comment by mikeurbach** [comment]
> I dunno. My thinking was that it is common for there to be PRs in flight for multiple versions of LLVM and it would be good if they didn't thrash each other.

Got it. For us, we still only test Python bindings in a nightly cron, so I guess this will be less of an issue for us in the short term. Anyway, it's an experiment, so looking forward to learning what works well from each other.

**Comment by mikeurbach** [comment]
@stellaraccident should the BUILD_SHARED_LIBS=ON setup still work? I hadn't tested it until I saw our CI failure, but I can reproduce that. It seems there is something unrelated to Python having visibility issues (for example, just a regular old C++ class generated from ODS not having a public member function visible). I'm digging in but maybe the visibility preset to hidden shouldn't be set when BUILD_SHARED_LIBS=ON? Or maybe there is something with how our libraries are declared in CMake I should tweak? We're just using add_mlir_library.

**Comment by mikeurbach** [comment]
Looks like setting `CMAKE_CXX_VISIBILITY_PRESET=default` in conjunction with `BUILD_SHARED_LIBS=ON` resolves the issue for our regular non-Python tests. It seems there's still something with `BUILD_SHARED_LIBS=ON` that doesn't work with the Python setup as I've configured it, but I'll dig into that too.

**Comment by stellaraccident** [comment]
> Looks like setting `CMAKE_CXX_VISIBILITY_PRESET=default` in conjunction with `BUILD_SHARED_LIBS=ON` resolves the issue for our regular non-Python tests. It seems there's still something with `BUILD_SHARED_LIBS=ON` that doesn't work with the Python setup as I've configured it, but I'll dig into that too.

Yes, hidden visibility is incompatible with shared library code that has not been specifically designed for it. The MLIR CAPI has been so designed but the c++ code has not. So hidden visibility is only valid for static builds (which have a shared library built for the CAPI). setting hidden visibility is good for deployment (as it will be better optimized) but it's not required.

I did code for BUILD_SHARED_LIBS=ON for the portion API, even though it isn't the best way. May have a big, though. At a guess, I'd point the finger at something wrong with the rpath in the upstream path, which has a special case

**Comment by mikeurbach** [comment]
Thanks, that makes sense. Until the same care goes into the C++ code as the C API, I'll only set the hidden visibility in static builds. It sounds like the BUILD_SHARED_LIBS flag only gets used in development, so I think that will be ok for now.

I'll check the Python API tomorrow and see what I find.

**Comment by mikeurbach** [comment]
@stellaraccident FYI the BUILD_SHARED_LIBS option is working. I needed to add some dependencies to PRIVATE_LINK_LIBS. We shouldn't be depending on some of those directly, but the upstream CAPIs appear to be missing (e.g. mlir::parseSourceFile, SymbolTable::lookup). Patching those dependencies in via PRIVATE_LINK_LIBS for now seems to be working.

**Comment by teqdruid** [comment]
When I try to set up a unified build on windows (on this branch), I get:

```
CMake Error at CMakeLists.txt:42 (message):
  CIRCT Python bindings require a unified build.  See docs/PythonBindings.md.


-- Configuring incomplete, errors occurred!
```

I'm using cmake 3.19.3 with the following command:

```
cmake ../llvm/llvm `
  -G "Visual Studio 16 2019" `
  -DLLVM_ENABLE_PROJECTS=mlir `
  -DCMAKE_BUILD_TYPE=Debug -Thost=x64 -DCMAKE_BUILD_TYPE=Debug `
  -DLLVM_ENABLE_ASSERTIONS=ON `
  -DLLVM_EXTERNAL_PROJECTS=circt `
  -DLLVM_EXTERNAL_CIRCT_SOURCE_DIR=.. `
  -DMLIR_ENABLE_BINDINGS_PYTHON=ON `
  -DCIRCT_BINDINGS_PYTHON_ENABLED=ON
```

**Comment by teqdruid** [comment]
Nevermind... Stupid powershell issue. Had to wrap the .. in quotes.

**Comment by mikeurbach** [comment]
Ah ok good. Is this branch working for you @teqdruid? The nightly tests finally pass on this branch: https://github.com/llvm/circt/actions/runs/1076700431. If you're good with this I'll merge it.

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,7 +4,7 @@
 ##
 ##===----------------------------------------------------------------------===//
 
-cmake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.13.4)
```
Comment (teqdruid) [inline CMakeLists.txt:7]: Do you need to bump this? How old is 3.13.4? Is it a version that most circt users would have?

**Review by teqdruid — APPROVED** [review]
You know my thoughts on cmake and build system changes: "if it works, ship it!"

File: CMakeLists.txt (hunk 2 above)
Comment (mikeurbach) [reply CMakeLists.txt:7]: I took this from npcomp, but I'm not sure exactly why this version was used. 3.13 was released in Nov 2018, but I think 3.10 was what is available by default on ubuntu 18.04. I'll see, if it's not required I won't bump it.

Comment (mikeurbach) [reply CMakeLists.txt:7]: Actually I guess this version comes from upstream MLIR: https://github.com/llvm/llvm-project/blob/9e9599ef782384847e734dfdb89d34a4bb8c6710/mlir/CMakeLists.txt#L6. It's also the documented version for building LLVM, so I think it's generally safe to assume CIRCT developers will have this: https://llvm.org/docs/GettingStarted.html#software.

**Comment by teqdruid** [comment]
Thanks!

**Pull request merged**